libfreerdp/utils/smartcard_call.c

```diff
@@ -43,23 +43,96 @@
+	do                                                           \
+	{                                                            \
+		if (ctx->emulated)                                       \
+		{                                                        \
+			_ret = Emulate_##fkt(ctx->emulation, ##__VA_ARGS__); \
+		}                                                        \
+		else                                                     \
+		{                                                        \
+			_ret = fkt(__VA_ARGS__);                             \
+		}                                                        \
+	} while (0)
+
+#define WRAP_WINSCARD_CALL_NORET(ctx, fkt, ...)           \
+	do                                                    \
+	{                                                     \
+		if (ctx->emulated)                                \
+		{                                                 \
+			Emulate_##fkt(ctx->emulation, ##__VA_ARGS__); \
+		}                                                 \
+		else                                              \
+		{                                                 \
+			fkt(__VA_ARGS__);                             \
+		}                                                 \
+	} while (0)
+#else // no smartcard emulation
+#define WRAP_WINSCARD_CALL_RET(ctx, fkt, _ret, ...) \
+	do                                              \
+	{                                               \
+		if (ctx->emulated)                          \
+		{                                           \
+			_ret = SCARD_F_INTERNAL_ERROR;          \
+		}                                           \
+		else                                        \
+		{                                           \
+			_ret = fkt(__VA_ARGS__);                \
+		}                                           \
+	} while (0)
+
+#define WRAP_WINSCARD_CALL_LONG(ctx, fkt, _ret, ...) \
+	do                                               \
+	{                                                \
+		if (ctx->emulated)                           \
+		{                                            \
+			_ret = 0;                                \
+		}                                            \
+		else                                         \
+		{                                            \
+			_ret = fkt(__VA_ARGS__);                 \
+		}                                            \
+	} while (0)
+
+#define WRAP_WINSCARD_CALL_NORET(ctx, fkt, ...) \
+	do                                          \
+	{                                           \
+		if (ctx->emulated)                      \
+		{                                       \
+		}                                       \
+		else                                    \
+		{                                       \
+			fkt(__VA_ARGS__);                   \
+		}                                       \
+	} while (0)
 #endif
 
 #define SCARD_MAX_TIMEOUT 60000
 
 struct s_scard_call_context
 {
+	bool emulated;
```

Comment:
* while most likely working, add `#include <stdbool.h>` or change to `BOOL`
* the variable name can easily be confused with`emulation` defined below, better use a more distinct name for this or the context below